contracts/VestingEscrow.vy

```diff
@@ -0,0 +1,348 @@
+    @param recipient Address to vest tokens for
+    @param start_time Epoch time at which token distribution starts
+    @param end_time Time until everything should be vested
+    @param cliff_length Duration after which the first portion vests
+    @param voting_adapter_addr VotingAdapter address
+    """
+    assert not self.initialized, "can only initialize once"
+    self.initialized = True
+
+    self.token = ERC20(token)
+    self.start_time = start_time
+    self.end_time = end_time
+    self.cliff_length = cliff_length
+    self.recipient = recipient
+    self.disabled_at = end_time  # Set to maximum time
+    self.voting_adapter_addr = voting_adapter_addr
+
+    return True
+
+
+@external
+@nonreentrant("lock")
+def activate(
+    amount: uint256,
+    owner: address,
+    manager: address,
+) -> bool:
+    """
+    @notice Fund and activate the contract. Requires amount of the token to be approved to vesting address
+    @dev This function is separate from `initialize` because we need to separate vesting creation and funing.
+         It may be called only once.
+    @param amount Address of the ERC20 token being distributed
+    @param owner Address of the vesting owner
+    @param manager Address of the vesting manager
+    """
+    assert not self.activated, "can only activate once"
+    self.activated = True
+
+    assert owner != empty(address), "zero_address owner"
+    self.owner = owner
+    self.manager = manager
+
+    assert self.token.transferFrom(
+        msg.sender, self, amount
+    ), "could not fund escrow"
+
+    self.total_locked = amount
+
+    log Activated(self.recipient, amount, owner, manager)
+
+    return True
+
+
+@external
+@view
+def get_token() -> address:
+    """
+    @notice Get vesting token address
+    """
+    return self.token.address
+
+
+@internal
+@view
+def _total_vested_at(time: uint256 = block.timestamp) -> uint256:
+    start: uint256 = self.start_time
+    end: uint256 = self.end_time
+    locked: uint256 = self.total_locked
+    if time < start + self.cliff_length:
+        return 0
+    return min(locked * (time - start) / (end - start), locked)
+
+
+@internal
+@view
+def _unclaimed(time: uint256 = block.timestamp) -> uint256:
+    return self._total_vested_at(time) - self.total_claimed
+
+
+@external
+@view
+def unclaimed() -> uint256:
+    """
+    @notice Get the number of unclaimed, vested tokens for recipient
+    """
+    # NOTE: if `revoke_unvested` is activated, limit by the activation timestamp
+    return self._unclaimed(min(block.timestamp, self.disabled_at))
+
+
+@internal
+@view
+def _locked(time: uint256 = block.timestamp) -> uint256:
+    return self.total_locked - self._total_vested_at(time)
+
+
+@external
+@view
+def locked() -> uint256:
+    """
+    @notice Get the number of locked tokens for recipient
+    """
+    # NOTE: if `revoke_unvested` is activated, limit by the activation timestamp
+    return self._locked(min(block.timestamp, self.disabled_at))
```

Comment:
Limit to `balance(self)` as in FullyRevokable.

Reply:
Done